runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/retries/impl/StandardRetryStrategy.kt

```diff
@@ -0,0 +1,160 @@
+ * @param tokenBucket The token bucket instance. Utilizing an existing token bucket will share call capacity between
+ * those scopes.
+ * @param backoffDelayer A delayer that can back off after the initial try to spread out the retries.
+ */
+class StandardRetryStrategy(
+    val options: StandardRetryStrategyOptions,
+    private val tokenBucket: RetryTokenBucket,
+    private val backoffDelayer: BackoffDelayer,
+) : RetryStrategy {
+    /**
+     * Retry the given block of code until it's successful. Note this method throws exceptions for non-successful
+     * outcomes from retrying.
+     */
+    override suspend fun <R> retry(policy: RetryPolicy<R>, block: suspend () -> R): R =
+        withTimeout(options.maxTimeMs) {
+            doTryLoop(block, policy, 1, tokenBucket.acquireToken(), null)
+        }
+
+    /**
+     * Perform a single iteration of the try loop. Execute the block of code, evaluate the result, and take action to
+     * terminate or enter the next iteration of the try loop.
+     * @param block The block of code to retry.
+     * @param policy The [RetryPolicy] to use for evaluating the execution result.
+     * @param attempt The ordinal index of the retry loop (starting at 1).
+     * @param fromToken A [RetryToken] which grants the strategy capacity to execute a try. This token is resolved
+     * inside the function by calling [notifySuccess][RetryToken.notifySuccess],
+     * [notifyFailure][RetryToken.notifyFailure], or [scheduleRetry][RetryToken.scheduleRetry].
+     * @param previousResult The [Result] from the prior loop iteration. This is used in the case of a timeout to
+     * include in the thrown exception.
+     */
+    private tailrec suspend fun <R> doTryLoop(
+        block: suspend () -> R,
+        policy: RetryPolicy<R>,
+        attempt: Int,
+        fromToken: RetryToken,
+        previousResult: Result<R>?,
+    ): R {
+        val callResult = runCatching { block() }
+        if (callResult.exceptionOrNull() is TimeoutCancellationException) {
+            throwTimeOut(fromToken, attempt, previousResult)
+        }
+
+        val nextToken = try {
+            when (val evaluation = policy.evaluate(callResult)) {
+                is RetryDirective.TerminateAndSucceed ->
+                    return success(fromToken, callResult)
+
+                is RetryDirective.TerminateAndFail ->
+                    throwFailure(fromToken, attempt, callResult)
+
+                is RetryDirective.RetryError ->
+                    if (attempt >= options.maxAttempts) {
+                        throwTooManyAttempts(fromToken, attempt, callResult)
+                    } else {
+                        // Prep for another loop
+                        backoffDelayer.backoff(attempt)
+                        fromToken.scheduleRetry(evaluation.reason)
+                    }
+            }
+        } catch (e: TimeoutCancellationException) {
```

Comment:
fix

We probably need to catch any exception that might happen from the backoff delayer / token retry such that we always call `token.notifyFailure()`. For instance if the coroutine `Job` is cancelled you won't get a `TimeoutCancellationException`, you would just get `CancellationException` and we would fail to call `notifyFailure()`